L63-dotnet-grpc-web.md

```diff
@@ -0,0 +1,76 @@
+
+Calling gRPC-Web is typically done using the JavaScript gRPC-Web library. This works for JavaScript solutions, but the modern .NET ecosystem allows for .NET applications to run in the browser using Web Assembly. A gRPC-Web .NET client is needed for this scenario.
+
+The grpc-dotnet gRPC server is built on ASP.NET Core and hosted on Kestrel, a full-featured, multi-purpose HTTP/1.1 and HTTP/2 server. ASP.NET Core has a [request pipeline that an application can plug middleware into](https://docs.microsoft.com/aspnet/core/fundamentals/middleware) to process the request and response. ASP.NET Core also supports configuring CORS for an application.
+
+The grpc-dotnet gRPC client is built on HttpClient, a full-featured, multi-purpose HTTP/1.1 and HTTP/2 client. HttpClient has an [outgoing request pipe that an application can plug handlers into](https://docs.microsoft.com/dotnet/api/system.net.http.httpmessagehandler).
+
+## Proposal
+
+* Add `Grpc.AspNetCore.Web`, a new package that acts as a in-process proxy for gRPC-Web. This package provides middleware via `UseGrpcWeb()` that detects gRPC-Web requests by their content-type, and translates them to gRPC HTTP/2. Request headers and the request message(s) are translated inbound, and response headers/footers and response message(s) are translated back to gRPC-Web outbound. Base64 content is encoded and decoded as content is read. There is configuration to control which services support gRPC-Web. A developer can choose to enable gRPC-Web for all services in `AddGrpcWeb`, or opt-in and opt-out on individual services.
+
+*Startup.cs* configuration file:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddGrpc();
+
+    // Enable gRPC-Web for all services
+    services.AddGrpcWeb(o => o.GrpcWebEnabled = true);
+}
+
+public void Configure(IApplicationBuilder app)
+{
+    app.UseRouting();
+
+    // Add gRPC-Web middleware
+    app.UseGrpcWeb();
+
+    app.UseEndpoints(endpoints =>
+    {
+        endpoints.MapGrpcService<GreeterService>();
+    });
+}
+```
+
+* Add `Grpc.Net.Client.Web`, a new package that acts as a gRPC-Web outgoing proxy for the .NET gRPC client. This package provides a `DelegatingHandler` that translates outgoing gRPC HTTP/2 requests to gRPC-Web. Base64 content is encoded and decoded as content is read. When a gRPC channel is created the developer can choose to configure it to use a HttpClient that has the `GrpcWebHandler`. When the handler is created the developer chooses whether to use `application/grpc-web` or `application/grpc-web-text`.
+
+```csharp
+// Create delegating handler and configure to use application/grpc-web-text
+var grpcWebHandler = new GrpcWebHandler(GrpcWebMode.GrpcWebText, new HttpClientHandler());
+
+var channel = GrpcChannel.ForAddress("https://localhost:5001", new GrpcChannelOptions
+    {
+        HttpClient = new HttpClient(grpcWebHandler)
+    });
+var client = Greeter.GreeterClient(channel);
+```
+
+### Interoperability
```

Comment:
@stanley-cheung can you confirm that https://github.com/johanbrandhorst/grpc-web-compatibility-test is a community-provided test suite and (while very useful) is not the "official" solution for testing grpc-web interoperability?
It would be good to have grpc-web interop infrastructure in place so we can track regressions in different implementations.

Reply:
As discussed during meeting, there is official test suite.